compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributeForall.cpp

```
@@ -87,9 +88,12 @@ LogicalResult resolveGPUMappedForallOp(RewriterBase &rewriter,
   assert(!(hasThreadMapping && hasWarpMapping));
   Value flatId = linearThreadId;
   if (hasWarpMapping) {
-    OpFoldResult subgroupSizeVal = rewriter.getIndexAttr(subgroupSize);
-    flatId = affine::makeComposedAffineApply(rewriter, loc, d0.floorDiv(d1),
-                                             {flatId, subgroupSizeVal});
+    flatId = rewriter
+                 .create<affine::AffineDelinearizeIndexOp>(
+                     loc, flatId,
+                     ArrayRef<int64_t>{flatWorkgroupSize / subgroupSize,
```

Can we add something like
```
if (flatWorkgroupSize % subgroupSize != 0) {
  forallOp->emitOpError("found warp mapped forall with non-multiple workgroup size");
  return failure();
}
```
I just realized this could silently pass through even if they aren't a multiple, and I'm fairly certain we don't handle such a case properly.

---

Done